rust-tdshim/src/mp.rs

```diff
@@ -157,55 +159,53 @@ fn parallel_accept_memory(apic_id: u64) {
     }
 }
 
+// Wait for APs to arrive by checking if they are available
 fn wait_for_ap_arrive(cpu_num: u32) {
     let mut mail_box = MailBox::new(get_mem_slice_mut(SliceType::MailBox));
-    log::info!("Waiting for APs to arrive...\n");
-    loop {
-        if mail_box.cpu_arrival() == cpu_num - 1 {
-            log::info!("All APs has arrived\n");
-            break;
-        }
-    }
-    mail_box.set_cpu_exiting(cpu_num);
-}
-
-fn wait_for_ap_exit(cpu_num: u32) {
-    let mail_box = MailBox::new(get_mem_slice_mut(SliceType::MailBox));
-    log::info!("Waiting for APs to exit...\n");
-    loop {
-        let cpu_exit = mail_box.cpu_exiting();
-
-        if cpu_exit == 1 {
-            log::info!("All APs has exited\n");
-            break;
+    for i in 1..cpu_num {
+        mail_box.set_apic_id(i);
```

Comment:
What about APIC ID 0? Is that invalid? If so, shouldn't `set_apic_id()` check this and return a `Result<()>`? Either way, a comment explaining why this is 1-based and not zero-based would help imho.

Reply:
@jodh-intel Thanks for your suggestions. The 0 is bootstrap processor (BSP) which is running this code to wakeup other vCPUs. I will add a comment to explain this.